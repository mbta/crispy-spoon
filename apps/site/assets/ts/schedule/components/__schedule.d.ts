import { TypedRoutes } from "../../stop/components/__stop";
import { Route } from "../../__v3api";

export interface SchedulePageData {
  connections: TypedRoutes[];
  pdfs: SchedulePDF[];
  teasers: string | null;
  hours: string;
  fares: Fare[];
  fare_link: string;
  holidays: Holiday[];
  route: Route;
  schedule_note: ScheduleNote | null;
  stops: SimpleStop[];
}

export interface ScheduleNote {
  peak_service: string;
  offpeak_service: string;
  exceptions: ServiceException[];
}

export interface ServiceException {
  type: string;
  service: string;
}

export interface SimpleStop {
  id: string;
  name: string;
}

export interface SchedulePDF {
  title: string;
  url: string;
}

export interface Fare {
  title: string;
  price: string;
}

export interface Holiday {
  name: string;
  date: string;
}
