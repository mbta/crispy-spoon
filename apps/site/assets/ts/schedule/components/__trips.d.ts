import { Route, Trip, Stop, PredictedOrScheduledTime } from "../../__v3api";

export interface Journey {
  trip: Trip;
  route: Route;
  departure: Departure;
}

export interface EnhancedJourney extends Journey {
  realtime: PredictedOrScheduledTime;
}

export interface Departure {
  time: string;
  schedule: Schedule | null;
  prediction: Prediction | null;
}

export interface TripDeparture {
  schedule: TripSchedule;
  prediction: Prediction | null;
}

export interface TripInfo {
  times: TripDeparture[];
  vehicle: Vehicle | null;
  vehicle_stop_name: string;
  stop_count: number;
  status: string;
  duration: number;
  fare: Fare;
}

export interface Prediction {
  delay: number;
  status: string | null;
  track: string | null;
}

export interface Schedule {
  time: string;
  stop_sequence: number;
  pickup_type: number;
  "flag?": boolean;
  "early_departure?": boolean;
  "last_stop?": boolean;
}

export interface TripSchedule extends Schedule {
  stop: Stop;
  fare: Fare;
}

export interface Vehicle {
  trip_id: string;
  stop_id: string;
  status: string;
}

export interface Fare {
  price: string;
  fare_link: string;
}
