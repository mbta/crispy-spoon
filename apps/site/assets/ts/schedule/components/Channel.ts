import { Socket, Channel } from "phoenix";

declare global {
  interface Window {
    channels: { [id: string]: Channel };
    socket?: Socket;
  }
}

type UpdateEventName = "reset" | "add" | "update" | "setChannel";

interface UpdateEvent<T> {
  event: UpdateEventName;
  data: T;
}

interface RemoveEvent {
  event: "remove";
  data: string[];
}

export type SocketEvent<T> = UpdateEvent<T> | RemoveEvent;

export const doInitChannel = (id: string): Channel => {
  if (!window.socket) {
    window.socket = new Socket("/socket", {});
    window.socket.connect();
  }

  if (!window.channels) {
    window.channels = {};
  }

  if (window.channels[id]) return window.channels[id];

  const channel = window.socket.channel(id, {});
  window.channels[id] = channel;

  channel
    .join()
    .receive("error", ({ reason }) =>
      // eslint-disable-next-line no-console
      console.error(`failed to join ${id}`, reason)
    )
    // eslint-disable-next-line no-console
    .receive("timeout", () => console.error(`timeout joining ${id}`))
    .receive("ok", () => {
      if (id !== "vehicles:remove") {
        const route = id.split(":");
        channel.push("init", {
          // eslint-disable-next-line camelcase
          route_id: route[1],
          // eslint-disable-next-line camelcase
          direction_id: route[2]
        });
      }
    });

  return channel;
};

export const initChannel = <T>(
  channelId: string,
  onData: (data: SocketEvent<T>) => void
): void => {
  const channel = doInitChannel(channelId);
  channel.on("data", onData);
};

// NB: this removes *all* handlers from the channel. We're currently throwing
// away the reference returned by `channel.on`, so we can't remove a specific
// handler, but we don't have a use case for that yet.
export const stopChannel = (id: string): void => {
  if (window.channels && window.channels[id]) {
    window.channels[id].off("data");
  }
};
