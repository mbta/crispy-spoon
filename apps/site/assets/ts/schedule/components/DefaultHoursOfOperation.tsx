import React, { ReactElement } from "react";
import ExpandableBlock from "../../components/ExpandableBlock";

const DefaultHoursOfOperation = ({
  hours
}: {
  hours: string;
}): ReactElement<HTMLElement> | null =>
  hours ? (
    <ExpandableBlock
      header={{ text: "Hours of Operation", iconSvgText: null }}
      initiallyExpanded={false}
      id="hours"
    >
      <div
        className="m-schedule-page__sidebar-hours"
        // eslint-disable-next-line react/no-danger
        dangerouslySetInnerHTML={{ __html: hours }}
      />
    </ExpandableBlock>
  ) : null;

export default DefaultHoursOfOperation;
