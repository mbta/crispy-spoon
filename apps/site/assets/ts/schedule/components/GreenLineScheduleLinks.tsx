import React, { ReactElement } from "react";
import ExpandableBlock from "../../components/ExpandableBlock";
import renderFa from "../../helpers/render-fa";
import { SchedulePDF } from "./__schedule";

const pdfLink = (pdf: SchedulePDF): ReactElement<HTMLElement> => (
  <a key={pdf.url} href={pdf.url} rel="noopener noreferrer" target="_blank">
    Open subway schedule PDF {renderFa("fa-arrow-up-right-from-square", "")}
  </a>
);

const getGreenLineScheduleLinks = (): ReactElement<HTMLElement> => {
  return (
    <>
      <div>
        <a href="/schedules/Green-B/line">B Line Schedule</a>
      </div>
      <div>
        <a href="/schedules/Green-C/line">C Line Schedule</a>
      </div>
      <div>
        <a href="/schedules/Green-D/line">D Line Schedule</a>
      </div>
      <div>
        <a href="/schedules/Green-E/line">E Line Schedule</a>
      </div>
    </>
  );
};

const GreenLineScheduleLinks = ({
  pdfs
}: {
  pdfs: SchedulePDF[];
}): ReactElement<HTMLElement> | null => {
  console.log(pdfs);
  return (
    <>
      <ExpandableBlock
        header={{ text: "Weekday Schedule", iconSvgText: null }}
        initiallyExpanded={false}
        id="weekday-hours"
      >
        <div className="m-schedule-page__sidebar-hours">
          <div className="fs-14 pt-14">Trains every 6-12 minutes</div>
          {getGreenLineScheduleLinks()}
        </div>
      </ExpandableBlock>
      <ExpandableBlock
        header={{ text: "Weekend Schedule", iconSvgText: null }}
        initiallyExpanded={false}
        id="weekend-hours"
      >
        <div className="m-schedule-page__sidebar-hours">
          <div className="font-weight-bold fs-18 pb-14">Saturday</div>
          <div className="fs-14 pt-14">Trains every 6-12 minutes</div>
          {getGreenLineScheduleLinks()}
          <div
            style={{
              borderBottomWidth: "0.5px",
              borderBottomColor: "black",
              borderBottomStyle: "solid"
            }}
            className="pt-18"
          ></div>
          <div className="font-weight-bold fs-18 pt-18 pb-14">Sunday</div>
          {getGreenLineScheduleLinks()}
          {<div className="fs-14 pt-18">{pdfLink(pdfs[0])}</div>}
        </div>
      </ExpandableBlock>
    </>
  );
};

export default GreenLineScheduleLinks;
