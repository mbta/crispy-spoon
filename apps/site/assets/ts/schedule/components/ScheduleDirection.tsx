import React, { ReactElement } from "react";
import { DirectionId, EnhancedRoute } from "../../__v3api";
import { ShapesById, RoutePatternsByDirection } from "./__schedule";

export interface Props {
  route: EnhancedRoute;
  directionId: DirectionId;
  shapesById: ShapesById;
  routePatternsByDirection: RoutePatternsByDirection;
}

interface ScheduleDirectionMenuProps {
  routePatternsByDirection: RoutePatternsByDirection;
  directionId: DirectionId;
}

const ScheduleDirectionMenu = ({
  routePatternsByDirection,
  directionId
}: ScheduleDirectionMenuProps): ReactElement<HTMLElement> => (
  <div>
    Direction Menu: {routePatternsByDirection[directionId].length} items for
    direction {directionId}
  </div>
);

const ScheduleDirection = ({
  route,
  directionId,
  shapesById,
  routePatternsByDirection
}: Props): ReactElement<HTMLElement> => (
  <div>
    <h5>Schedule Direction Component</h5>
    <p>{route.direction_names[directionId]}</p>
    <p>
      active shape:{" "}
      {
        shapesById[
          routePatternsByDirection[directionId].slice(0, 1)[0].shape_id
        ].name
      }
    </p>
    <ScheduleDirectionMenu
      directionId={directionId}
      routePatternsByDirection={routePatternsByDirection}
    />
  </div>
);

export default ScheduleDirection;
