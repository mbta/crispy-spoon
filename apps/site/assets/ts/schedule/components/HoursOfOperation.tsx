import React, { ReactElement } from "react";
import { isGreenLine, isRapidTransit } from "../../models/route";
import { EnhancedRoute } from "../../__v3api";
import DefaultHoursOfOperation from "./DefaultHoursOfOperation";
import GreenLineScheduleLinks from "./GreenLineScheduleLinks";
import RapidTransitHoursOfOperation from "./RapidTransitHoursOfOperation";
import { SchedulePDF, ScheduleNote } from "./__schedule";

const HoursOfOperation = ({
  route,
  pdfs,
  hours,
  scheduleNote
}: {
  route: EnhancedRoute;
  pdfs: SchedulePDF[];
  hours: string;
  scheduleNote: ScheduleNote | null;
}): ReactElement<HTMLElement> | null => {
  if (isGreenLine(route)) {
    return <GreenLineScheduleLinks pdfs={pdfs} />;
  }
  if (isRapidTransit(route)) {
    return (
      <RapidTransitHoursOfOperation
        pdfs={pdfs}
        route={route}
        scheduleNote={scheduleNote}
      />
    );
  }
  return <DefaultHoursOfOperation hours={hours} />;
};

export default HoursOfOperation;
