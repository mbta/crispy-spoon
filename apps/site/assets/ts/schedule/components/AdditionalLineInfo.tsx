import React, { ReactElement } from "react";
import { Holiday, SchedulePDF, Fare, ScheduleNote } from "./__schedule";
import { EnhancedRoute } from "../../__v3api";
import { TypedRoutes } from "../../stop/components/__stop";
import PDFSchedules from "./PDFSchedules";
import Connections from "./Connections";
import ContentTeasers from "./ContentTeasers";
import Fares from "./Fares";
import UpcomingHolidays from "./UpcomingHolidays";
import HoursOfOperation from "./HoursOfOperation";
import { isSubwayRoute } from "../../models/route";

interface Props {
  teasers: string | null;
  pdfs: SchedulePDF[];
  connections: TypedRoutes[];
  fares: Fare[];
  fareLink: string;
  route: EnhancedRoute;
  holidays: Holiday[];
  scheduleNote: ScheduleNote | null;
  hours: string;
}

const AdditionalLineInfo = ({
  connections,
  pdfs,
  teasers,
  fares,
  holidays,
  fareLink,
  route,
  hours,
  scheduleNote
}: Props): ReactElement<HTMLElement> => (
  <>
    <ContentTeasers teasers={teasers} />
    <PDFSchedules pdfs={pdfs} />
    <Connections connections={connections} />
    <Fares fares={fares} fareLink={fareLink} routeType={route.type} />
    {/* Only show the hours for non subway lines */}
    {!isSubwayRoute(route) && (
      <HoursOfOperation
        route={route}
        pdfs={pdfs}
        scheduleNote={scheduleNote}
        hours={hours}
      />
    )}
    <UpcomingHolidays holidays={holidays} />
  </>
);

export default AdditionalLineInfo;
