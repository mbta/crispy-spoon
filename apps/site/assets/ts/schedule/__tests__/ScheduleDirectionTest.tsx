import React from "react";
import renderer from "react-test-renderer";
import { createReactRoot } from "../../app/helpers/testUtils";
import ScheduleDirection from "../components/ScheduleDirection";
import { EnhancedRoute } from "../../__v3api";
import { RoutePatternsByDirection, ShapesById } from "../components/__schedule";

const route = {
  type: 3,
  name: "SL2",
  long_name: "Design Center - South Station",
  id: "742",
  direction_names: {
    0: "Outbound",
    1: "Inbound"
  },
  direction_destinations: {
    0: "Design Center",
    1: "South Station"
  },
  description: "key_bus_route",
  "custom_route?": false,
  header: "",
  alert_count: 0
} as EnhancedRoute;
const directionId = 1;
const routePatternsByDirection = {
  "0": [
    {
      typicality: 1,
      time_desc: null,
      shape_id: "7420039",
      route_id: "742",
      representative_trip_id: "40606000",
      name: "Drydock",
      id: "742-1-0",
      direction_id: 0
    }
  ],
  "1": [
    {
      typicality: 1,
      time_desc: null,
      shape_id: "7420037",
      route_id: "742",
      representative_trip_id: "40606147",
      name: "South Station",
      id: "742-_-1",
      direction_id: 1
    }
  ]
} as RoutePatternsByDirection;
const shapesById = {
  "7420037": {
    stop_ids: ["31255"],
    priority: 3,
    polyline: "xyz",
    name: "South Station",
    id: "7420037",
    direction_id: 1
  }
} as ShapesById;

it("it renders", () => {
  createReactRoot();
  const tree = renderer
    .create(
      <ScheduleDirection
        route={route}
        directionId={directionId}
        routePatternsByDirection={routePatternsByDirection}
        shapesById={shapesById}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
