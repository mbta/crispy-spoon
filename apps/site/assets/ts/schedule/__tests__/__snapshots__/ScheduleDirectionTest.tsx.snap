// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<div>
  <h5>
    Schedule Direction Component
  </h5>
  <p>
    Inbound
  </p>
  <p>
    active shape:
     
    South Station
  </p>
  <div>
    Direction Menu: 
    1
     items for direction 
    1
  </div>
</div>
`;
